pkg/activator/throttler_test.go

```diff
@@ -502,3 +502,70 @@ func endpointsSubset(hostsPerSubset, subsets int) []corev1.EndpointSubset {
 	}
 	return resp
 }
+
+func TestInfiniteBreaker(t *testing.T) {
+	b := &infiniteBreaker{
+		broadcast: make(chan struct{}),
+	}
+
+	// Verify initial condition.
+	if got, want := b.Capacity(), 0; got != want {
+		t.Errorf("Cap=%d, want: %d", got, want)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	if b.Maybe(ctx, nil) {
+		t.Error("Should have failed, but didn't")
+	}
+
+	b.UpdateConcurrency(1)
+	if got, want := b.Capacity(), 1; got != want {
+		t.Errorf("Cap=%d, want: %d", got, want)
+	}
+
+	// Verify we call the thunk when we have achieved capacity.
+	// Twice.
+	for i := 0; i < 2; i++ {
+		ctx, cancel = context.WithCancel(context.Background())
+		cancel()
+		res := false
+		if !b.Maybe(ctx, func() { res = true }) {
+			t.Error("Should have succeeded, but didn't")
+		}
+		if !res {
+			t.Error("thunk was not invoked")
+		}
+	}
+
+	// Scale to zero
+	b.UpdateConcurrency(0)
+
+	// Repeat initial test.
+	ctx, cancel = context.WithCancel(context.Background())
+	cancel()
+	if b.Maybe(ctx, nil) {
+		t.Error("Should have failed, but didn't")
+	}
+	if got, want := b.Capacity(), 0; got != want {
+		t.Errorf("Cap=%d, want: %d", got, want)
+	}
+
+	// And now do the async test.
+	ctx, cancel = context.WithTimeout(context.Background(), time.Millisecond*50)
```

Comment:
This might be a source of intermittent failures. Do we need the timeout?

Reply:
Changed to with cancel.